--- src/crypto/tls/common.go
+++ src/crypto/tls/common.go
@@ -62,7 +62,7 @@
 	maxCiphertext      = 16384 + 2048 // maximum ciphertext payload length
 	maxCiphertextTLS13 = 16384 + 256  // maximum ciphertext length in TLS 1.3
 	recordHeaderLen    = 5            // record header length
-	maxHandshake       = 65536        // maximum handshake we support (protocol max is 16 MB)
+	maxHandshake       = 262144       // maximum handshake we support (protocol max is 16 MB)
 	maxUselessRecords  = 16           // maximum number of consecutive non-advancing records
 )
 
