From d1f8df825632c244e4c087c7be83abe37d9ff9a6 Mon Sep 17 00:00:00 2001
From: Moritz Sanft <58110325+msanft@users.noreply.github.com>
Date: Fri, 29 Aug 2025 17:06:48 +0200
Subject: [PATCH] drop .osrel UKI section

Signed-off-by: Moritz Sanft <58110325+msanft@users.noreply.github.com>
---
 mkosi/__init__.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mkosi/__init__.py b/mkosi/__init__.py
index 65cac772..0266a595 100644
--- a/mkosi/__init__.py
+++ b/mkosi/__init__.py
@@ -1706,7 +1706,7 @@ def build_uki(
     json_out = False
 
     arguments: list[PathString] = [
-        "--os-release", f"@{workdir(context.root / 'usr/lib/os-release')}",
+        # "--os-release", f"@{workdir(context.root / 'usr/lib/os-release')}",
         "--uname", kver,
         "--linux", workdir(kimg),
         *flatten(["--join-profile", os.fspath(workdir(profile))] for profile in profiles),
-- 
2.51.0

